dali/pipeline/executor/executor.cc

```diff
@@ -524,6 +530,12 @@ void Executor<WorkspacePolicy, QueuePolicy>::RunHelper(OpNode &op_node, Workspac
       in.SetLayout({});
     }
   }
+
+  // If it is the end of an epoch, create a checkpoint
+  if (checkpointing_ && (iteration_id + 1) % checkpointing_epoch_size_ == 0) {
+    auto &cpt = GetCurrentIterationData(iteration_id + 1).checkpoint.GetOpCheckpoint(op_node.id);
+    op_node.op->SaveState(cpt, ws.has_stream() ? std::optional{ws.stream()} : std::nullopt);
+  }
```

Comment:
As a side note, you can consider saving some information in the checkpoint here, that for this iteration you actually saved a checkpoint. When the user calls `GetCurrentCheckpoint` we can verify that we actually have something to get, and the checkpoint was obtained in the correct place/time.